### PR TITLE
Add module support to csnex

### PR DESCRIPTION
**Pull request by gitlarryf**

### CODE REVIEW

File: exec/csnex/Support.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+{
+    public class Support
+    {
+        static public List<string> Paths;
+
+        public Support(string source_path)
+        {
+            Paths = new List<string>();
+            // Always check current directory;
+            Paths.Add(string.Format(".{0}", Path.DirectorySeparatorChar));
```
Comment (ghewgill) [inline exec/csnex/Support.cs:15]: This seems significantly different from the code in support.cpp. The order of paths should be the same as what is already established. Your code changes the order here with no information about why.

File: exec/csnex/Support.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            // Also check the Neon executable path.
+            Paths.Add(string.Format("{0}{1}", source_path, Path.DirectorySeparatorChar));
+            // Then check the cwd.
+            Paths.Add(Environment.CurrentDirectory);
```
Comment (ghewgill) [inline exec/csnex/Support.cs:19]: What is the purpose of this? You have already added the current directory above.

File: exec/csnex/Support.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                string[] evpaths = ne.Split(Path.PathSeparator);
+                foreach (string s in evpaths) {
+                    // Don't bother pushing empty paths.
+                    if (s != null && s.Length > 0) {
+                        Paths.Add(s);
+                    }
+                }
+            }
+            // Next, check for .neonpath, and process any paths that might exist in it.
+            if (File.Exists(".neonpath")) {
```
Comment (ghewgill) [inline exec/csnex/Support.cs:33]: This sort of existence check is not robust code, it is a potential race condition. It's possible that the file could be removed by something else between the time you check the existence and try to open it. Instead, simply try to open the file and handle any failure to open in the usual way.

File: exec/csnex/Support.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                StreamReader sr = new StreamReader(".neonpath");
+                while (!sr.EndOfStream) {
```
Comment (ghewgill) [inline exec/csnex/Support.cs:35]: This kind of "end of file" checking is also not robust. The `ReadLine` function already tells you when you are at the end of the file (it returns `null`). Use that instead.

File: exec/csnex/Support.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            // Then check the cwd.
+            Paths.Add(Environment.CurrentDirectory);
+
+            // Get paths from %NEONPATH% environment variable, if there is one.
+            string ne = Environment.GetEnvironmentVariable("NEONPATH");
+            if (ne != null) {
+                string[] evpaths = ne.Split(Path.PathSeparator);
+                foreach (string s in evpaths) {
+                    // Don't bother pushing empty paths.
+                    if (s != null && s.Length > 0) {
```
Comment (ghewgill) [inline exec/csnex/Support.cs:27]: Under what conditions would `Split()` return a `null` in the array? I don't see that mentioned anywhere in the documentation. You do not need to handle the `null` case if it is not documented to happen.

**Review by ghewgill — CHANGES_REQUESTED** [review]

File: exec/csnex/Exec.cs [hunk 6]
```diff
@@ -26,13 +27,29 @@ public ActivationFrame(int nest, ActivationFrame outer, int size, int stack_dept
             }
         }
 
-        public Executor(Bytecode bc)
+        sealed class StackFrame
```
[the remark this answers is not in the document]
Comment (gitlarryf) [reply exec/csnex/Exec.cs:30]: Renamed to CallStack, to match cnex naming convention.

File: exec/csnex/Exec.cs [hunk 7]
```diff
@@ -791,17 +841,38 @@ void CALLP()
...
+            int mod = Bytecode.Get_VInt(module.Bytecode.code, ref ip);
+            int fun = Bytecode.Get_VInt(module.Bytecode.code, ref ip);
+            int efi = 0;
+            if (callstack.Count >= param_recursion_limit) {
+                Raise("StackOverflowException", "");
+                return;
+            }
+
+            Module m = modules[module.Bytecode.strtable[mod]];
+            //Module m = LocateModule(module.Bytecode.strtable[mod]);
```
[the remark this answers is not in the document]
Comment (gitlarryf) [reply exec/csnex/Exec.cs:864]: Removed.

File: exec/csnex/Support.cs (hunk 1 above)
Comment (gitlarryf) [reply exec/csnex/Support.cs:15]: Changed order of the load paths.

File: exec/csnex/Support.cs (hunk 2 above)
Comment (gitlarryf) [reply exec/csnex/Support.cs:19]: Removed.

File: exec/csnex/Support.cs (hunk 3 above)
Comment (gitlarryf) [reply exec/csnex/Support.cs:33]: Use a try/catch for this now.

File: exec/csnex/Support.cs (hunk 4 above)
Comment (gitlarryf) [reply exec/csnex/Support.cs:35]: Removed sr.EndOfStream check.

File: exec/csnex/Support.cs (hunk 5 above)
Comment (gitlarryf) [reply exec/csnex/Support.cs:27]: Removed != null check.

Comment (ghewgill) [reply exec/csnex/Support.cs:28]: The check still seems to be there?

File: exec/csnex/Support.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                }
+            }
+            // Next, check for .neonpath, and process any paths that might exist in it.
+            try {
+                StreamReader sr = new StreamReader(".neonpath");
+                string line = null;
+                do {
+                    line = sr.ReadLine();
+                    // Don't bother pushing empty paths.
+                    if (line != null && line.Length > 0) {
```
Comment (ghewgill) [inline exec/csnex/Support.cs:40]: You've got confusing logic here and a double check for null. Check once for null and use break to get out of the loop.

File: exec/csnex/Support.cs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                        Paths.Add(line);
+                    }
+                } while (line != null);
+            } catch {
```
Comment (ghewgill) [inline exec/csnex/Support.cs:44]: It is inappropriate to catch and ignore *all* exceptions.

**Review by ghewgill — CHANGES_REQUESTED** [review]

**Pull request merged**